### PR TITLE
[Feature] CAIN Net

### PR DESCRIPTION
**Pull request by Yshuo-Li**

### CODE REVIEW

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+    b, c, h, w = x.shape
+    if up:
+        y = F.pixel_shuffle(x, scale)
+    else:
+        h = int(h / scale)
+        w = int(w / scale)
+        x = x.view(b, c, h, scale, w, scale)
+        shuffle_out = x.permute(0, 1, 3, 5, 2, 4).contiguous()
+        y = shuffle_out.view(b, c * scale * scale, h, w)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:31]: How about we put this as a separate function, and store it in `mmediting/mmedit/models/common/`? That could be useful for other methods. 

Also, we usually call this "pixel_unshuffle". So it's better to put it in a separate function.

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    if down >= h or right >= w:
+        function = nn.ReplicationPad2d
+    else:
+        function = nn.ReflectionPad2d
+    padding_input = function(padding=[left, right, up, down])
+    padding_output = function(padding=[0 - left, 0 - right, 0 - up, 0 - down])
+    return padding_input, padding_output
+
+
+class ConvNorm(nn.Module):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:69]: How about `ConvNormWithReflectionPad` ? I think the reflection padding should be mentioned in the class name.

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        self.reflection_pad = nn.ReflectionPad2d(reflection_padding)
+        self.conv = nn.Conv2d(
+            in_channels,
+            out_channels,
+            stride=stride,
+            kernel_size=kernel_size,
+            bias=True)
+
+        self.norm = norm
+        if norm == 'IN':
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:99]: `norm.lower() == 'in'`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+            self.norm = nn.InstanceNorm2d(
+                out_channels, track_running_stats=True)
+        elif norm == 'BN':
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:102]: `elif norm.lower() == 'BN':`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+            self.norm = nn.BatchNorm2d(out_channels)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:103]: Should add one more condition: 
```
elif norm is not None:
  raise ValueError(...)

```

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        """
+
+        out = self.reflection_pad(x)
+        out = self.conv(out)
+        if self.norm:
+            out = self.norm(out)
+        return out
+
+
+class CALayer(nn.Module):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:122]: Suggest to use full name: `ChannelAttentionLayer`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+    Args:
+        mid_channels (int): Channel number of the intermediate features.
+        reduction (int): Channel reduction of CA. Default: 16.
+    """
+
+    def __init__(self, mid_channels, reduction=16):
+        super().__init__()
+
+        # global average pooling: feature --> point
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:133]: `# global average pooling: (n, c, h, w) --> (n, c, 1, 1)`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        self.avg_pool = nn.AdaptiveAvgPool2d(1)
+        # feature channel downscale and upscale --> channel weight
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:135]: Not so sure what this means (I understand the operation below, but the comment needs to be modified)

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+            Tensor: Output tensor with shape (n, c, h, w).
+            Tensor: CA tensor with shape (n, c, 1, 1).
+        """
+
+        y = self.avg_pool(x)
+        y = self.channel_attention(y)
+        return x * y, y
+
+
+class RCA(nn.Module):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:166]: Use `ResidualChannelAttention`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    """Residual Channel Attention Module.
+
+    Args:
+        mid_channels (int): Channel number of the intermediate features.
+        kernel_size (int): Kernel size of RCA. Default: 3.
+        downscale (bool): Down scale or not. Default: False. Default: False.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:172]: `downscale (bool): Down scale or not. Default: False.`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        reduction (int): Channel reduction of CA. Default: 16.
+        return_ca (bool): Return CA tensor or not. Default: False.
+        norm (None | function): Norm layer. If None, no norm layer.
+            Default: None.
+        act (function): activate function. Default: nn.ReLU(True).
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:177]: activation function

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+        out += res
+
+        if self.return_ca:
+            return out, ca
+        else:
+            return out
+
+
+class ResidualGroup(nn.Module):
+    """Residual Group.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:234]: Residual Group, consisting of a stack of residual channel attention, followed by a convolution.

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+
+    Args:
+        block_layer (nn.module): nn.module class for basic block.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:237]: nn.Module

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+            kernel_size=3,
+            reduction=reduction,
+            act=act)
+        self.conv_last = nn.Conv2d(mid_channels, mid_channels, 3, 1, 1)
+
+    def forward(self, xs, padding_flag=False):
+        """Forward function.
+
+        Args:
+            xs (Tensor): Input tensor with shape (n, 2, c, h, w).
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:335]: better not to use `xs` to represent the images. Maybe use `imgs`?

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+            padding_flag (bool): Padding or not. Default: False.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:336]: Where is this argument checked?

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+        function = nn.ReplicationPad2d
+    else:
+        function = nn.ReflectionPad2d
+    padding_input = function(padding=[left, right, up, down])
+    padding_output = function(padding=[0 - left, 0 - right, 0 - up, 0 - down])
+    return padding_input, padding_output
+
+
+class ConvNorm(nn.Module):
+    """ReflectionPad2d, Conv2d (and Norm).
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:70]: Apply reflection padding, followed by a convolution, with an optional normalization.

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+
+    Args:
+        in_channels (int): Channel number of input features.
+        out_channels (int): Channel number of input features.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:74]: output features

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+        kernel_size (int): Kernel size of convolution layer.
+        stride (int): Stride of size of convolution layer. Default: 1.
+        norm (None | function): Norm layer. If None, no norm layer.
+            Default: None.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:78]: `norm (str | None): Normalization layer. If it is None, no normalization is performed. Default: None.`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 19]
```diff
@@ -0,0 +1,386 @@
...
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_size,
+                 stride=1,
+                 norm=None):
+        super().__init__()
+
+        reflection_padding = kernel_size // 2
+        self.reflection_pad = nn.ReflectionPad2d(reflection_padding)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:90]: `self.reflection_pad = nn.ReflectionPad2d(kernel_size // 2)` is fine.

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+        self.conv = nn.Conv2d(
+            in_channels,
+            out_channels,
+            stride=stride,
+            kernel_size=kernel_size,
+            bias=True)
+
+        self.norm = norm
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:98]: `self.norm = None`

**Review by ckkelvinchan — COMMENTED** [review]
@Yshuo-Li  Here are some comments. Some of the codes are not tested~

File: mmedit/models/backbones/sr_backbones/cain_net.py (hunk 15 above)
Comment (Yshuo-Li) [reply mmedit/models/backbones/sr_backbones/cain_net.py:336]: CAIN model, in another PR

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+        followed by a convolution.
+
+    Args:
+        block_layer (nn.Module): nn.module class for basic block.
+        num_block_layers (int): number of blocks.
+        mid_channels (int): Channel number of the intermediate features.
+        kernel_size (int): Kernel size of ResidualGroup.
+        reduction (int): Channel reduction of CA. Default: 16.
+        act (function): activate function. Default: nn.ReLU(True).
+        norm (None | function): Norm layer. If None, no norm layer.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:247]:
```
norm (str | None): Normalization layer. If it is None, no
  normalization is performed. Default: None.
```

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+    """
+
+    def __init__(self,
+                 block_layer,
+                 num_block_layers,
+                 mid_channels,
+                 kernel_size,
+                 reduction,
+                 act,
+                 norm=False):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:258]: `norm=None`

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+
+    Args:
+        in_channels (int): Channel number of inputs. Default: 3.
+        out_channels (int): Channel number of outputs. Default: 3.
+        kernel_size (int): Kernel size of CAINNet. Default: 3.
+        num_block_groups (int): Number of block groups. Default: 5.
+        num_block_layers (int): Number of blocks in a group. Default: 12.
+        depth (int): Down scale depth, scale = 2**depth. Default: 3.
+        reduction (int): Channel reduction of CA. Default: 16.
+        norm (None | function): Norm layer. If None, no norm layer.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:302]:
```
norm (str | None): Normalization layer. If it is None, no
  normalization is performed. Default: None.
```

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+
+    def __init__(self,
+                 in_channels=3,
+                 out_channels=3,
+                 kernel_size=3,
+                 num_block_groups=5,
+                 num_block_layers=12,
+                 depth=3,
+                 reduction=16,
+                 padding=7,
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:315]: add to docstring

File: mmedit/models/backbones/sr_backbones/cain_net.py [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+        in_channels (int): Channel number of inputs. Default: 3.
+        out_channels (int): Channel number of outputs. Default: 3.
+        kernel_size (int): Kernel size of CAINNet. Default: 3.
+        num_block_groups (int): Number of block groups. Default: 5.
+        num_block_layers (int): Number of blocks in a group. Default: 12.
+        depth (int): Down scale depth, scale = 2**depth. Default: 3.
+        reduction (int): Channel reduction of CA. Default: 16.
+        norm (None | function): Norm layer. If None, no norm layer.
+            Default: None.
+        act (function): activate function. Default: nn.LeakyReLU(0.2, True)).
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/cain_net.py:304]: The default activation function is not the same as in other modules. Is it intentional?

**Review by ckkelvinchan — COMMENTED** [review]
Maybe we should put this model into another folder, say `interpolation_backbone`?

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 26]
```diff
@@ -0,0 +1,343 @@
...
+        norm (None | function): Norm layer. If None, no norm layer.
+            Default: None.
+        act (function): activation function. Default: nn.LeakyReLU(0.2, True).
+    """
+
+    def __init__(self,
+                 mid_channels,
+                 kernel_size=3,
+                 reduction=16,
+                 norm=False,
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:160]: `norm=None`

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 27]
```diff
@@ -0,0 +1,343 @@
...
+                 num_block_groups=5,
+                 num_block_layers=12,
+                 depth=3,
+                 reduction=16,
+                 norm=None,
+                 padding=7,
+                 act=nn.LeakyReLU(0.2, True)):
+        super().__init__()
+
+        mid_channels = in_channels * pow(4, depth)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:271]: `mid_channels = in_channels * (4**depth)` should be slightly faster.

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 28]
```diff
@@ -0,0 +1,343 @@
...
+        self.scale = pow(2, depth)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:272]: `self.scale = 2**depth` should be slightly faster.

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 29]
```diff
@@ -0,0 +1,343 @@
...
+    padding_width, padding_height = 0, 0
+    if w != ((w >> padding) << padding):
+        padding_width = (((w >> padding) + 1) << padding) - w
+    if h != ((h >> padding) << padding):
+        padding_height = (((h >> padding) + 1) << padding) - h
+    left, right = padding_width // 2, padding_width - padding_width // 2
+    up, down = padding_height // 2, padding_height - padding_height // 2
+    # print(up, down, left, right)
+    if down >= h or right >= w:
+        function = nn.ReplicationPad2d
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:36]: Not covered by unittest.

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 30]
```diff
@@ -0,0 +1,343 @@
...
+
+class ConvNormWithReflectionPad(nn.Module):
+    """Apply reflection padding, followed by a convolution,
+        which can be followed by an optional normalization.
+
+    Args:
+        in_channels (int): Channel number of input features.
+        out_channels (int): Channel number of output features.
+        kernel_size (int): Kernel size of convolution layer.
+        stride (int): Stride of size of convolution layer. Default: 1.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:53]: Can consider removing this, as it seems that stride is always to 1.

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 31]
```diff
@@ -0,0 +1,343 @@
...
+        mean1 = x1.mean(2, keepdim=True).mean(3, keepdim=True)
+        mean2 = x2.mean(2, keepdim=True).mean(3, keepdim=True)
+        x1 -= mean1
+        x2 -= mean2
+
+        if padding_flag:
+            padding_function, depadding_function = get_padding_functions(
+                x1, self.padding)
+            x1 = padding_function(x1)
+            x2 = padding_function(x2)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:311]: Is there a case when `self.padding > 0` but `padding_flag=False`?

Comment (Yshuo-Li) [reply mmedit/models/backbones/video_interpolation_backbones/cain_net.py:311]: It will not pad in this case.

Comment (ckkelvinchan) [reply mmedit/models/backbones/video_interpolation_backbones/cain_net.py:311]: So can use only `self.padding` to determine? 

Specifically, how is `padding_flag` set? If `padding_flag` is set based on the size of the input, I guess only `self.padding` should be sufficient?

Comment (Yshuo-Li) [reply mmedit/models/backbones/video_interpolation_backbones/cain_net.py:311]: `padding_flag` is necessary in #691

File: mmedit/models/backbones/video_interpolation_backbones/cain_net.py [hunk 32]
```diff
@@ -0,0 +1,336 @@
...
+from mmedit.models.registry import BACKBONES
+from mmedit.utils import get_root_logger
+
+
+def get_padding_functions(x, padding=7):
+    """Generate padding function for CAIN.
+
+    This function produces two functions to pad and depad a tensor, given the
+    number of pixels to be padded. When applying padding and depadding
+    sequentially, the original tensor is obtained.
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/video_interpolation_backbones/cain_net.py:16]: As discussed, add docstring here.

**Review by ckkelvinchan — APPROVED** [review]

**Pull request merged**